Bug #5 fish passes out of fishtank

Akashae98 opened the pull request:
# 5
a ver como lo ves, he creado una clase para boundingbox.
Te explico por que he creado el atributo heightfishtail, como el objeto se dibuja a partir del ovalo del cuerpo del pez, osea la posición (0,0) en el centro y el borde izquierdo, quería dibujar los métodos proporcionales para que fuera reutilizable el código. El problema que la cola que sale por debajo añade un extra al bottom. Si luego cogiera una imagen y la posicion a partir de la que dibujar el boundingbox , para calcular el top y el bottom solo necesitaría la mitad del alto. Pero para este caso he añadido un margen de la cola para que se tenga en cuenta un bottom mas largo. Como este se define en el constructor si la imagen es proporcional y el punto es desde el centro se resuelve pasandole un zero a ese heightfishtail y ya estaría, pero tu como lo ves?
Y ahora hay otro problema jaja, que como los peces se añaden random por toda la pecera cuando entran mas allá de la zona de rebote se quedan rebotando por el borde sin poder salir.

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into Feature-Red_fish_realistic_drawing

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into Feature-Red_fish_realistic_drawing

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into Feature-Red_fish_realistic_drawing

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into Feature-Red_fish_realistic_drawing

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into Feature-Red_fish_realistic_drawing

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
clase abstracta modificada

Commit:
cambios abstracta

Commit:
Merge origin/Feature-Red_fish_realistic_drawing into #13_Abstract_class_Animation_(to_draw)

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Commit:
clase abstracta modificada
cambio nombre del main
x e y fuera de la abstracta, se pasan al metodo draw

Commit:
changes in constructor of fish, new class AnimationBubbleidle,
new class fo random color..
falta integrar bubble con color

Commit:
new class AnimationBubbleidle,
new class fo random color,
method randomPoint in Fishtank

Commit:
more bubbles

Commit:
more bubbles

Commit:
cambios en rebound, fishtank, fish, size..

Commit:
cambios getrandompoint

Commit:
cambios getrandompoint

Commit:
cambios Color en constructor fish, randomcolor se usa al instanciar Fish

Commit:
nombre Position para objeto pair

Commit:
nombre Position ora objeto pair

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into bug_#5_fish_passes_out_of_fishtank

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java
#	src/main/java/com/mycompany/animacionpecera/Fish.java
#	src/main/java/com/mycompany/animacionpecera/FishTank.java

Commit:
Merge #13_Abstract_class_Animation_(to_draw) into bug_#5_fish_passes_out_of_fishtank

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java
#	src/main/java/com/mycompany/animacionpecera/Fish.java
#	src/main/java/com/mycompany/animacionpecera/FishTank.java

Commit:
new class position

Commit:
new class position

Commit:
new class position

Akashae98 committed:
cambios en metodos movimiento por objeto position

Commit:
Merge origin/#13_Abstract_class_Animation_(to_draw) into #13_Abstract_class_Animation_(to_draw)

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java
#	src/main/java/com/mycompany/animacionpecera/Fish.java
#	src/main/java/com/mycompany/animacionpecera/FishTank.java
#	src/main/java/com/mycompany/animacionpecera/Movement.java

Commit:
Objeto position en constructores

Commit:
cambios reboun, nueva clase boundinbox

Commit:
cambios en rebound, nueva clase boundingbox

Commit:
Merge origin/bug_#5_fish_passes_out_of_fishtank into bug_#5_fish_passes_out_of_fishtank

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java
#	src/main/java/com/mycompany/animacionpecera/FishTank.java
#	src/main/java/com/mycompany/animacionpecera/MainScene.java
#	src/main/java/com/mycompany/animacionpecera/Movement.java

Commit:
valores rebote ajuste

Commit:
Boundingbox simplified

Commit:
Boundingbox simplified, comments

Commit:
Boundingbox simplified, comments

Commit:
Boundingbox simplified, comments

Commit:
Boundingbox simplified, comments

Commit:
comments

Commit:
added variable heightFishTail

gcervifi commented:
Primero de todo, esta rama tiene conflictos. No se que merge has hecho, pero hay que arreglarlo. 

Segundo, el BoundingBox consiste de 4 puntos, o de un punto inicial junto con el alto y el ancho para obtener los otros 3 puntos.
BoundingBox no sabe de colas o scale, solo sabe 4 puntos. Puedes incluir metodos de 
- isInside(Point point): bool
para que te sea mas facil hacer los calculos en otros sitios.


Fish no puede generar su BounddingBox, porque depende de la parte de animacion para que lo defina. Crea un metodo en Animacion que sea obtener el BoundingBox. Toda la logica especifaca para caluclar el box final de cada pez, metelo dento de animacion correspondiente. 

La idea, de momento (ya veremos que neceistamos cambiar) es:
- Crea pez con una posicion. 
- Obten el boundingBox de la animacion.
- Mueve el pez en funcion de movimiento y su boundingBox
- Pinta el pez.

De momento ignora el bug del pez en los bordes, sera algo que mejor ver en la issue  de movimiento.

Commit:
limite no añadir en bordes pecera

Commit:
limite no añadir en bordes pecera

Akashae98 committed:
Merge branch 'master' into bug_#5_fish_passes_out_of_fishtank

Commit:
method getBoundingBox in Animation,
integrated with fishiddle

Commit:
changes boundingbox, 4 points
changed method boundingbox

Commit:
Merge origin/bug_#5_fish_passes_out_of_fishtank into bug_#5_fish_passes_out_of_fishtank

# Conflicts:
#	src/main/java/com/mycompany/animacionpecera/Fish.java

Akashae98 commented:
Bueno de momento lo he hecho abstracto, creo que he hecho todos los cambios como decías, a ver qué tal ahora

Akashae98 commented:
Okey, con punto te refieres a posición (x,y) ? Y sería un punto por cada
lado del box?

El jue, 26 jun 2025, 16:05, gcervifi ***@***.***> escribió:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/mycompany/animacionpecera/BoundingBox.java
> <https://github.com/Akashae98/pecera/pull/16#discussion_r2169161348>:
>
> > + */
> +package com.mycompany.animacionpecera;
> +
> +/*
> + * BoundingBox represents the area occupied by a fish.
> + * It calculates the fish's real position and dimensions
> + * based on its size and position.
> + */
> +public class BoundingBox {
> +
> +    private double left;
> +    private double top;
> +    private double right;
> +    private double bottom;
> +
> +    public BoundingBox(double left, double top, double right, double bottom) {
>
> Bounding box necesitas 4 puntos. Un punto esta compuesto por 2 variables.
> Aqui no hay ningun punto definido.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Akashae98/pecera/pull/16#pullrequestreview-2962265686>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKS42RNWUOHTOLWLUCALJID3FP435AVCNFSM6AAAAACAD66PDGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSNRSGI3DKNRYGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Akashae98 commented:
Vale no, son las esquinas en teoría no?

El jue, 26 jun 2025, 18:27, Carol Vilar ***@***.***> escribió:

> Okey, con punto te refieres a posición (x,y) ? Y sería un punto por cada
> lado del box?
>
> El jue, 26 jun 2025, 16:05, gcervifi ***@***.***> escribió:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In src/main/java/com/mycompany/animacionpecera/BoundingBox.java
>> <https://github.com/Akashae98/pecera/pull/16#discussion_r2169161348>:
>>
>> > + */
>> +package com.mycompany.animacionpecera;
>> +
>> +/*
>> + * BoundingBox represents the area occupied by a fish.
>> + * It calculates the fish's real position and dimensions
>> + * based on its size and position.
>> + */
>> +public class BoundingBox {
>> +
>> +    private double left;
>> +    private double top;
>> +    private double right;
>> +    private double bottom;
>> +
>> +    public BoundingBox(double left, double top, double right, double bottom) {
>>
>> Bounding box necesitas 4 puntos. Un punto esta compuesto por 2 variables.
>> Aqui no hay ningun punto definido.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Akashae98/pecera/pull/16#pullrequestreview-2962265686>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BKS42RNWUOHTOLWLUCALJID3FP435AVCNFSM6AAAAACAD66PDGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSNRSGI3DKNRYGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

Commit:
points in boundingbox

Akashae98 commented:
de momento he hecho el de 4 puntos, a ver si ahora esta bien

gcervifi commented on src/main/java/com/mycompany/animacionpecera/BoundingBox.java:
```diff
@@ -0,0 +1,67 @@
+    public double getBottom() {
+        return bottomRight.y;
+    }
+    // Checks if a given point is inside this bounding box.
```
Que significa left right bottom y encima que sea un double? Como sabes si cada uno es x o y? Borra estas funciones y usa los metodos de postion.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Fish.java:
```diff
@@ -19,31 +19,31 @@ public class Fish {
     }
 
     //Method of movement
     public void move(int width, int height) {
```
Borra los parametros width y height si no haces nada con ellos

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java:
```diff
@@ -13,17 +12,52 @@
+    @Override
+    public BoundingBox getBoundingBox(Position position) {
+        //the base of the bodyfish 
+        double bodyWidth = 38 * size;
```
De donde sacas estos valores? No hay ningun 38 en otra parte del codigo. Y si son numeros que se reutilizan, ponlo como una constante para que todos puedan acceder.

Commit:
cambios pr

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Bubble.java:
```diff
@@ -29,16 +28,16 @@ public Bubble(Position position, double size, double speed) {
         //the position of y decreases to the top
-        y -= speed;
-        this.position = new Position(x, y);
+        position.y -= speed;
```
Asi te aseguras que sin querer no modificas la posicion, por ejemplo.

Commit:
cambios en position: metodo para sumar o restar posicion + getters.
cambio en metodo rebound, el rebote invierte la dieccion de forma constante.
nuevo metodo teletransporte, controla que nunca aparezca un borde del box fuera

Commit:
coralfish sin atributo imagen

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Fish.java:
```diff
@@ -33,12 +33,16 @@ public Fish(Position pos, double size, Color color) {
+        boundingBox = anim.getBoundingBox(position);
+        //teletransport relocateds a fish inside canvas
+        position = movement.teletransport(boundingBox, position);
+
```
El bug del pez atrapado no se arregla haciendo un teleport. Se arregla fixeando el metodo displacement. En cambio, para la burbuja entiendo que si haces un teleport. Ahi si iria bien ese metodo.

gcervifi replied:
Si haces el rebound bien, los peces, Si aparcen fuera, navegaran al interior. Si solo te preocupa cuando aparecen los peces, que esten dentro de la pecera, puedes limitar la zona donde aparecen los peces, dejando un margen. Pero no me preocuparia por ahora. Recuerda, en las PR es solo para hacer la PR. No vayas metiendo otros fixes. Porque como vez, todo cambio puede añadir errores. Y hay que mirar entonces los primeros cambios y los segundos. CUando es mas facil subir uno y luego hacer otra PR.

gcervifi replied:
Ignora el teleport por el momento.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Movement.java:
```diff
@@ -14,29 +14,68 @@ public class Movement {
-    //this method changes direction before object passes the limits
+    /*this method changes direction before object passes the limits, 
+    always to the opposite direction*/
     public double[] rebound(BoundingBox box, double currentDx, double currentDy) {
```
Te estas complicando la vida. Primero, si quieres que siempre salga positivo, usa el metodo de valor absoluto, no me hagas cuadrados y divisiones. Segundo, solo estas usando dY y no dX. Si comparas con X tendras que usar Dx, si comparas con Y usa Dy. Tercero, new direction siempre es 1? dy²/dy² siempre sera 1. Fijate, al rebotar la primera vez los peces se vuelven locos.

No ibas por mal camino, antes tenias en un solo if la comprobacion de los dos lados y invertias. El fix es comprobar cada lado de la pecera independientemente. Luego, para cada lado, comprobar si sale del limite Y LUEGO, invertir solo en funcion de DX actual. Es decir, si el pez esta fuera, pero se esta moviendo HACIA la pecera, no modifiques DX.

Commit:
cambios pr

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationCoralFish.java:
```diff
@@ -6,37 +6,87 @@
-    public AnimationCoralFish(double size) {
+    private Color color;
+
+    public AnimationCoralFish(double size, Color color) {
```
Para que quieres Color si importas el archivo directamente?

gcervifi replied:
Para el bounding box? Entonces crealo tu en el constructor, no lo pases como parametro del constructor.

Akashae98 replied:
Uy es que estaba utilizando en el constructor de fish que si el color era coral poner la instancia de pez de coral y sino la de fishidle

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationFishIdle.java:
```diff
@@ -22,9 +22,7 @@ public AnimationFishIdle(double size, boolean hasFishFin, Color color) {
-    /*Note: the fish's actual position (0,0) refers to the center-left point of the object.
-    * Therefore, the bounding box must account for displacement toward only the top,
-    * right, and bottom to fully cover the fish's visual area.
+    /*Note: the fish's actual position (0,0) refers to a center-left point in the object.
```
No entiendo que hablas de (0,0) las cooredenadas son globales, y (0,0) significa la esquina superior izquierda del canvas.

Akashae98 replied:
sino le pongo que es el punto de referencia o el punto de posicion dentro del pez

gcervifi commented on src/main/java/com/mycompany/animacionpecera/Movement.java:
```diff
@@ -19,60 +19,50 @@ public class Movement {
+            newDirection[1] *= -1;
         }
 
         return newDirection;
```
Que newDirection sea un array de dx, dy, nos indica que en el futuro habra que cambiarlo por un objeto de Vector.

Commit:
cambios pr

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationCoralFish.java:
```diff
@@ -16,9 +16,9 @@ public class AnimationCoralFish extends Animation {
+    public AnimationCoralFish(double size) {
         super(size);
-        this.color = color;
+        this.color = Color.WHITE;
```
Si es una constante de la clase color, no hace falta ni que lo guardes en el constructor, usalo directamente donde lo necesites.

gcervifi commented:
El boundingBox de la burbuja esta mal.
![image](https://github.com/user-attachments/assets/3db08203-19ea-4b84-991d-639b7dd3b2ac)

Commit:
calculos boundingbox burbuja actualizado
metodo drawBoundingbox se hereda de clase animation

Commit:
calculos boundingbox burbuja actualizado
metodo drawBoundingbox se hereda de clase animation

Commit:
cambios pr

Commit:
cambios pr

Commit:
cambios pr

Commit:
Merge commit '52e3b9e9be648adf22dea35a5edb03790dea58c2' into bug_#5_fish_passes_out_of_fishtank

Akashae98 commented on src/main/java/com/mycompany/animacionpecera/AnimationCoralFish.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public BoundingBox getBoundingBox(Position position) {
```
no? donde te refieres? si estoy utilizando el size en el metodo de gc.drawImage le tengo que usar el size luego en el boundingbox no?

gcervifi replied:
Si el punto dado es la zona central, topLeft sera punto.x - Width/2, punto.y -Height/2 y asi para los cuatro.

Akashae98 replied:
sii, lo tenia hecho ya asi🥲

gcervifi commented:
Cambia ya al InteliJ.

Akashae98 commented:
> Cambia ya al InteliJ.

😭

Commit:
calculos boundingbox burbuja y coralfish actualizado
posicion en centro

Commit:
mismo color boundingbox

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationBubbleIdle.java:
```diff
@@ -54,7 +54,10 @@ public void draw(GraphicsContext gc, Position position) {
-        drawBoundingBox(gc, boundingBox, Color.WHITE);
+        drawBoundingBox(gc, boundingBox, Color.MAGENTA);
+
+        Position center = position.displacement(size / 2, size / 2);
```
Lo mismo aqui, no desplaces el punto.

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationCoralFish.java:
```diff
@@ -23,45 +23,45 @@ public AnimationCoralFish(double size) {
+
+        double width = image.getWidth() * size;
+        double height = image.getHeight() * size;
+        Position center = position.displacement(width / 2, height / 2);
```
Fuera

gcervifi commented on src/main/java/com/mycompany/animacionpecera/AnimationCoralFish.java:
```diff
@@ -23,45 +23,45 @@ public AnimationCoralFish(double size) {
         drawBoundingBox(gc, boundingBox, Color.WHITE);
-        gc.strokeText("o", pos.x, pos.y);
+
+        Position center = pos.displacement(width / 2, height / 2);
```
Fuera

Commit:
mismo color boundingbox

Akashae98 commented on src/main/java/com/mycompany/animacionpecera/AnimationBubbleIdle.java:
```diff
@@ -20,28 +20,28 @@ public AnimationBubbleIdle(double size) {
-        double extra_bottom = 0.25 * size;
+        double width = size;
+        double height = size;
+        Position center = position.displacement(width / 2, height / 2);
```
okey, aunque el metodo cree un nuevo punto?

Akashae98 replied:
quiero decir, igualmente tengo que sacar los valores de la posición dada para obtener el centro, no seria eso con una nueva posicion?

gcervifi commented:
![imagen](https://github.com/user-attachments/assets/47ffd41b-4db5-460b-861d-2d1d42b5e68e)
Caso 1 es lo que tenias al principio. Dibujabas a partir del punto.
Caso 2 estas ahora. A partir del punto, desplazas Width/2 y Height/2 Y LUEGO dibujas alrededor del punto.
Caso3 es lo que tienes que hacer. A partir del punto, dibuja alrededor.

Commit:
coralfish y burbuja se dibujan desde el centro

gcervifi commented:
Haz Squash and merge
